
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="http://arcadia-science.github.io/arcadia-computational-training/arcadia-users-group/20240416-intro-to-ml1/make_peptides_ml_dataset/">
      
      <link rel="icon" href="../../../assets/Arcadia_LogoOnly-XL-K.png">
      <meta name="generator" content="mkdocs-1.4.0, mkdocs-material-8.5.6">
    
    
      
        <title>Make peptides ml dataset - Arcadia Science Computational Training</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.20d9efc8.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.cbb835fc.min.css">
        
          
          
          <meta name="theme-color" content="#000000">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="black" data-md-color-accent="light-blue">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#documentation-for-how-the-dataset-was-made" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Arcadia Science Computational Training" class="md-header__button md-logo" aria-label="Arcadia Science Computational Training" data-md-component="logo">
      
  <img src="../../../assets/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Arcadia Science Computational Training
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Make peptides ml dataset
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/arcadia-science/arcadia-computational-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    arcadia-computational-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Arcadia Science Computational Training" class="md-nav__button md-logo" aria-label="Arcadia Science Computational Training" data-md-component="logo">
      
  <img src="../../../assets/logo.png" alt="logo">

    </a>
    Arcadia Science Computational Training
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/arcadia-science/arcadia-computational-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    arcadia-computational-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Computational Training Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../overview/" class="md-nav__link">
        Arcadia Users Group
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../workshops/overview/" class="md-nav__link">
        Workshops
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../../CONTRIBUTING/" class="md-nav__link">
        Contribute
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#documentation-for-how-the-dataset-was-made" class="md-nav__link">
    Documentation for how the dataset was made
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#documentation-of-data-set-contents" class="md-nav__link">
    Documentation of data set contents
  </a>
  
    <nav class="md-nav" aria-label="Documentation of data set contents">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#column-descriptions" class="md-nav__link">
    Column descriptions
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#potential-future-additions-or-improvements" class="md-nav__link">
    Potential future additions or improvements
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>Make peptides ml dataset</h1>

<h2 id="documentation-for-how-the-dataset-was-made">Documentation for how the dataset was made</h2>
<ol>
<li>We used the environment defined in <code>make_peptides_ml_dataset.yml</code> to make the data set below.</li>
<li>Originally, we downloaded the peptipedia database using the following Google Drive link:  https://drive.google.com/file/d/1x3yHNl8k5teHlBI2FMgl966o51s0T8i_/view?usp=sharing. The link for this database is provided by the <a href="https://github.com/ProteinEngineering-PESB2/Peptipedia2.0">peptipedia GitHub repo</a>. For persistence, we uploaded this file to <a href="https://osf.io/kb7z6/">this OSF repository</a>, as well as some of the processed data created by the code below.</li>
<li>Unzip the database.</li>
<li>The unzipped archive contains three files, <code>backup_sql.zip</code>, <code>peptipedia_csv.zip</code>, and <code>peptipedia_fasta.zip</code>.
   Unzipping <code>peptipedia_fasta.zip</code> results in a FASTA file <code>peptipedia.fasta</code> that contains all of the peptide sequences from the peptipedia database.
   Unzipping <code>peptipedia_csv.zip</code> results in a CSV file <code>peptipedia.csv</code> that records the peptide id, peptide sequence, and predicted bioactivity labels.
   Unzipping <code>backup_sql.zip</code> results in a postgres SQL dump file <code>backup_dump.sql</code>.
   This file contains metadata associated with the peptide sequences.
   To extract the metadata to separate CSV tables, we used the python code:
   <div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">re</span>

<span class="n">output_file_prefix</span> <span class="o">=</span> <span class="s1">&#39;table_&#39;</span>
<span class="n">current_table</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">output_file</span> <span class="o">=</span> <span class="kc">None</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;backup_dump.sql&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">dump_file</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">dump_file</span><span class="p">:</span>
        <span class="c1"># Check for COPY command start</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;COPY&#39;</span><span class="p">):</span>
            <span class="n">table_name</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>  <span class="c1"># Assumes table name is the second word</span>
            <span class="n">current_table</span> <span class="o">=</span> <span class="n">table_name</span>
            <span class="n">output_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">output_file_prefix</span><span class="si">}{</span><span class="n">current_table</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;\.&#39;</span><span class="p">):</span>
            <span class="c1"># End of data for the current table</span>
            <span class="k">if</span> <span class="n">output_file</span><span class="p">:</span>
                <span class="n">output_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
                <span class="n">output_file</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">elif</span> <span class="n">output_file</span><span class="p">:</span>
            <span class="c1"># Convert tab-separated values to comma-separated values</span>
            <span class="n">csv_line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;,&#39;</span><span class="p">)</span>
            <span class="n">output_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">csv_line</span><span class="p">)</span>
</code></pre></div></li>
</ol>
<p>This code extracts the following CSV files:
   <div class="highlight"><pre><span></span><code>table_public.activity.csv
table_public.activity_spectral.csv
table_public.db.csv
table_public.encoding.csv
table_public.gene_ontology.csv
table_public.index.csv
table_public.patent.csv
table_public.peptide.csv
table_public.peptide_has_activity.csv
table_public.peptide_has_db_has_index.csv
table_public.peptide_has_go.csv
table_public.pfam.csv
</code></pre></div></p>
<ol>
<li>
<p>Filter to peptides that are at least k in length. MMSeqs2 clustering in step 4 will fail if any of the sequences are below the k-mer length used for clustering.
   <div class="highlight"><pre><span></span><code>seqkit<span class="w"> </span>seq<span class="w"> </span>--min-len<span class="w"> </span><span class="m">10</span><span class="w"> </span>-o<span class="w"> </span>peptipedia_minlen10.fasta<span class="w"> </span>peptipedia.fasta
</code></pre></div></p>
</li>
<li>
<p>Cluster the sequences.
   The goal of this step is to reduce homology in peptide sequences so that the data set can be split in any way and not have pollution between the train and test sets.
   It uses a threshold of 80% because this has been previously shown to be sufficient for homology reduction in protein sequences (<a href="https://academic.oup.com/nar/article/47/8/e43/5314020">source</a>).
   <div class="highlight"><pre><span></span><code>mmseqs<span class="w"> </span>easy-cluster<span class="w"> </span>peptipedia_minlen10.fasta<span class="w"> </span>mmseqs/peptipedia_minlen10-0.8<span class="w"> </span>tmp<span class="w"> </span>--min-seq-id<span class="w"> </span><span class="m">0</span>.8
</code></pre></div></p>
</li>
<li>
<p>Filter to clusters composed of peptides that all have the same bioactivity prediction and write a TSV file.
   Many of the peptides that end up in clusters have multiple bioactivity predictions (e.g. antimicrobial and therapeutic).
   The script below filters to peptide clusters that all have at least one bioactivity label that is the same.
   It also adds additional metadata (GO, PFAM) from the SQL database to the data.frame of peptide information.
   <div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">janitor</span><span class="p">)</span>

<span class="c1"># The goal of this script is to create a demo machine learning data set from peptides and their metadata.</span>
<span class="c1"># It filters peptides in the peptipedia database to large clusters of similar peptides</span>
<span class="c1"># (80% identity or greater) that all have the same bioactivity label.</span>
<span class="c1"># It then joins selected peptides to their metadata.</span>

<span class="c1"># filter to peptides clusters with many members that have the same bioactivity --------</span>

<span class="c1"># read in cluster information from mmseqs2 linclust</span>
<span class="n">clusters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;~/Downloads/backup_peptipedia_29_03_2023/mmseqs/peptipedia_minlen10-0.8_cluster.tsv&quot;</span><span class="p">,</span>
<span class="w">                     </span><span class="n">col_names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;rep&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;member&quot;</span><span class="p">))</span>

<span class="c1"># read in peptide bioactivity information</span>
<span class="n">peptide_bioactivity</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_csv</span><span class="p">(</span><span class="s">&#39;~/Downloads/backup_peptipedia_29_03_2023/peptipedia.csv&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">clean_names</span><span class="p">()</span>

<span class="c1"># make the data set a little bit smaller by filtering to clusters that have &gt; 2 members</span>
<span class="c1"># otherwise, the lines of code to find clusters with all of the same bioactivity takes too long</span>
<span class="n">cluster_sizes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusters</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">rep</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">tally</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">arrange</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>

<span class="n">large_clusters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cluster_sizes</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="n">clusters_filtered</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusters</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">rep</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">large_clusters</span><span class="o">$</span><span class="n">rep</span><span class="p">)</span>

<span class="c1"># filter the bioactivity data set to match</span>
<span class="n">peptide_bioactivity_filtered</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_bioactivity</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">idpeptide</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">clusters_filtered</span><span class="o">$</span><span class="n">member</span><span class="p">)</span>

<span class="c1"># filter to clusters that all have the same bioactivity</span>
<span class="c1"># (e.g. all peptides have a predicted bioactivity of &quot;antimicrobial&quot; in a given cluster;</span>
<span class="c1">#  note that most peptides have multiple bioactivity labels.</span>
<span class="c1">#  this code filters to clusters where at least one bioactivity label is the same for all peptides in the cluster.)</span>
<span class="n">peptide_bioactivity_with_clusters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">left_join</span><span class="p">(</span><span class="n">clusters_filtered</span><span class="p">,</span><span class="w"> </span><span class="n">peptide_bioactivity_filtered</span><span class="p">,</span>
<span class="w">                                               </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;member&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;idpeptide&quot;</span><span class="p">))</span>

<span class="n">peptide_bioactivity_with_clusters_long</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_bioactivity_with_clusters</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">pivot_longer</span><span class="p">(</span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">starts_with</span><span class="p">(</span><span class="s">&quot;allergen&quot;</span><span class="p">)</span><span class="o">:</span><span class="nf">last_col</span><span class="p">(),</span><span class="w"> </span><span class="n">names_to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bioactivity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">values_to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;value&quot;</span><span class="p">)</span>

<span class="c1"># Find clusters where all members have a 1 for any bioactivity</span>
<span class="n">clusters_all_1_any_bioactivity</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_bioactivity_with_clusters_long</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">rep</span><span class="p">,</span><span class="w"> </span><span class="n">bioactivity</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">summarise</span><span class="p">(</span><span class="n">all_ones</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">all</span><span class="p">(</span><span class="n">value</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="n">.groups</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;drop&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">all_ones</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">distinct</span><span class="p">(</span><span class="n">rep</span><span class="p">,</span><span class="w"> </span><span class="n">.keep_all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># filter to those clusters where all members have at least 1 bioactivity the same</span>
<span class="n">peptide_group_bioactivities</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_bioactivity_with_clusters</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">rep</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">clusters_all_1_any_bioactivity</span><span class="o">$</span><span class="n">rep</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">left_join</span><span class="p">(</span><span class="n">clusters_all_1_any_bioactivity</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;rep&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="c1"># select only the representative for the cluster so that there can&#39;t be pollution between test and train data no matter how it is split;</span>
<span class="w">  </span><span class="c1"># sequences will at most be &lt; 80% similar to any other sequence in the dataset.</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">rep</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">member</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">select</span><span class="p">(</span><span class="n">peptipedia_peptide_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rep</span><span class="p">,</span><span class="w"> </span><span class="n">peptide_sequence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sequence</span><span class="p">,</span><span class="w"> </span><span class="n">group_bioactivity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bioactivity</span><span class="p">)</span>

<span class="c1"># remove peptides where there are very few representatives (&lt; 50)</span>
<span class="n">few_peptides_with_group_bioactivity</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_group_bioactivities</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">group_bioactivity</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">tally</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>

<span class="n">peptide_group_bioactivities_filtered</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_group_bioactivities</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="o">!</span><span class="n">group_bioactivity</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">few_peptides_with_group_bioactivity</span><span class="o">$</span><span class="n">group_bioactivity</span><span class="p">)</span>

<span class="c1"># Record all bioactivity predictions for each peptide by concatenating the bioactivity names</span>
<span class="n">peptide_bioactivities_concatenated</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_bioactivity_with_clusters_long</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">value</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">rep</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">member</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">select</span><span class="p">(</span><span class="n">peptipedia_peptide_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rep</span><span class="p">,</span><span class="w"> </span><span class="n">bioactivity</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">peptipedia_peptide_id</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">summarise</span><span class="p">(</span><span class="n">all_bioactivities</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">str_c</span><span class="p">(</span><span class="n">bioactivity</span><span class="p">,</span><span class="w"> </span><span class="n">collapse</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;;&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">.groups</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;drop&#39;</span><span class="p">)</span>


<span class="c1"># Join this back to your original (or relevant) dataset to add the concatenated bioactivities column</span>
<span class="n">final_dataset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">peptide_group_bioactivities_filtered</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">left_join</span><span class="p">(</span><span class="n">peptide_bioactivities_concatenated</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;peptipedia_peptide_id&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">distinct</span><span class="p">()</span>

<span class="c1"># join with other peptide metadata ----------------------------------------</span>

<span class="n">gene_ontology</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_csv</span><span class="p">(</span><span class="s">&quot;~/Downloads/backup_peptipedia_29_03_2023/table_public.gene_ontology.csv&quot;</span><span class="p">,</span>
<span class="w">                          </span><span class="n">col_names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;peptideid&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;go_id&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;go_description&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;go_category&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">select</span><span class="p">(</span><span class="n">peptideid</span><span class="p">,</span><span class="w"> </span><span class="n">go_id</span><span class="p">)</span>

<span class="n">pfam</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_csv</span><span class="p">(</span><span class="s">&quot;~/Downloads/backup_peptipedia_29_03_2023/table_public.pfam.csv&quot;</span><span class="p">,</span>
<span class="w">                 </span><span class="n">col_names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;peptideid&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;pfam_id&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;pfam_name&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;pfam_type&quot;</span><span class="p">))</span>

<span class="n">final_dataset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">final_dataset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">left_join</span><span class="p">(</span><span class="n">gene_ontology</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;peptipedia_peptide_id&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;peptideid&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">left_join</span><span class="p">(</span><span class="n">pfam</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;peptipedia_peptide_id&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;peptideid&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">distinct</span><span class="p">()</span>

<span class="n">final_dataset</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">group_by</span><span class="p">(</span><span class="n">group_bioactivity</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">tally</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">arrange</span><span class="p">(</span><span class="nf">desc</span><span class="p">(</span><span class="n">n</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">knitr</span><span class="o">::</span><span class="nf">kable</span><span class="p">()</span>

<span class="nf">write_tsv</span><span class="p">(</span><span class="n">final_dataset</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;~/Downloads/backup_peptipedia_29_03_2023/peptides_for_ml_tmp2.tsv&quot;</span><span class="p">)</span>
</code></pre></div></p>
</li>
<li>
<p>Convert the output TSV to a FASTA file that can be used to calculate peptide characteristics (molecular weight, etc).
   <div class="highlight"><pre><span></span><code>cut<span class="w"> </span>-f1,2<span class="w"> </span>peptides_for_ml_tmp.tsv<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-n<span class="w"> </span>+2<span class="w"> </span><span class="p">|</span><span class="w"> </span>seqkit<span class="w"> </span>tab2fx<span class="w"> </span>&gt;<span class="w"> </span>peptides_for_ml.fasta
</code></pre></div></p>
</li>
<li>
<p>Calculate peptide characteristics (length, mw, etc.)</p>
</li>
</ol>
<div class="highlight"><pre><span></span><code>curl<span class="w"> </span>https://raw.githubusercontent.com/Arcadia-Science/peptigate/main/scripts/characterize_peptides.py
python<span class="w"> </span>characterize_peptides.py<span class="w"> </span>peptides_for_ml.fasta<span class="w"> </span>peptides_for_ml_characterized.tsv
</code></pre></div>
<ol>
<li>Combine peptide metadata with other information</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="n">peptide_characterization</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;~/Downloads/backup_peptipedia_29_03_2023/peptides_for_ml_characterized.tsv&quot;</span><span class="p">)</span>

<span class="n">final_dataset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">final_dataset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">left_join</span><span class="p">(</span><span class="n">peptide_characterization</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;peptipedia_peptide_id&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;peptide_id&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">nchar</span><span class="p">(</span><span class="n">peptide_sequence</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">rename</span><span class="p">(</span><span class="n">bioactivity</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group_bioactivity</span><span class="p">)</span>

<span class="nf">write_tsv</span><span class="p">(</span><span class="n">final_dataset</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;peptides_ml_dataset.tsv&quot;</span><span class="p">)</span>
</code></pre></div>
<ol>
<li>Check that the data set is predictive of the label "group_bioactivity".
   <div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">caret</span><span class="p">)</span>

<span class="nf">set.seed</span><span class="p">(</span><span class="m">123</span><span class="p">)</span>

<span class="n">final_dataset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="s">&quot;peptides_ml_dataset.tsv&quot;</span><span class="p">)</span>
<span class="n">final_dataset_filtered</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">final_dataset</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">peptipedia_peptide_id</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">peptide_sequence</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">all_bioactivities</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">go_id</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">pfam_id</span><span class="p">,</span>
<span class="w">        </span><span class="o">-</span><span class="n">pfam_name</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">pfam_type</span><span class="p">)</span>
<span class="n">trainingIndex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">createDataPartition</span><span class="p">(</span><span class="n">final_dataset_filtered</span><span class="o">$</span><span class="n">bioactivity</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">.</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<span class="n">trainingData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">final_dataset_filtered</span><span class="p">[</span><span class="n">trainingIndex</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="n">testData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">final_dataset_filtered</span><span class="p">[</span><span class="o">-</span><span class="n">trainingIndex</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>

<span class="n">control</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">trainControl</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s">&quot;cv&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="o">=</span><span class="m">10</span><span class="p">)</span><span class="w"> </span><span class="c1"># 10-fold cross-validation</span>
<span class="n">model</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">train</span><span class="p">(</span><span class="n">bioactivity</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="o">=</span><span class="n">trainingData</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s">&quot;rf&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trControl</span><span class="o">=</span><span class="n">control</span><span class="p">)</span>

<span class="n">predictions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">predict</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="o">=</span><span class="n">testData</span><span class="p">)</span>
<span class="nf">confusionMatrix</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span><span class="w"> </span><span class="nf">as.factor</span><span class="p">(</span><span class="n">testData</span><span class="o">$</span><span class="n">bioactivity</span><span class="p">))</span>
</code></pre></div></li>
</ol>
<p>The <code>confusionMatrix()</code> command prints the following two results to the console.
   The model has the following accuracy:
   <div class="highlight"><pre><span></span><code>Overall Statistics

            Accuracy : 0.6188          
              95% CI : (0.5805, 0.6559)
 No Information Rate : 0.2496          
 P-Value [Acc &gt; NIR] : &lt; 2.2e-16       

               Kappa : 0.5316          

Mcnemar&#39;s Test P-Value : NA
</code></pre></div></p>
<p>and with per class performance of:
   <div class="highlight"><pre><span></span><code>                             Reference
Prediction                 amphibian_defense anti_gram antibacterial_antibiotic antimicrobial growth_factor hormone immunological_activity neurological_activity propeptide signal_peptide therapeutic
amphibian_defense                        0         0                        0             3             0       0                      0                     0          0              0           0
anti_gram                                0         0                        0             1             0       0                      0                     0          0              0           0
antibacterial_antibiotic                 0         0                        0             0             0       0                      0                     0          1              3           0
antimicrobial                           12         9                        9           119             0       9                      9                     4          3             16          11
growth_factor                            0         0                        0             0             0       0                      0                     0          0              0           0
hormone                                  0         0                        0             3             2       8                      1                     4          3              5           3
immunological_activity                   3         0                        0            10             0       3                     47                     3          1              1           2
neurological_activity                    0         0                        0             0             0       0                      1                     0          0              1           1
propeptide                               0         0                        1             9             1       1                      1                     1         59             10           2
signal_peptide                           1         0                        6             9             9      14                      2                     9         18            120           1
therapeutic                              1         1                        3            11             0       1                      0                     1          2              0          56


Statistics by Class:

                     Class: amphibian_defense Class: anti_gram Class: antibacterial_antibiotic Class: antimicrobial Class: growth_factor Class: hormone Class: immunological_activity Class: neurological_activity Class: propeptide Class: signal_peptide Class: therapeutic
Sensitivity                          0.000000         0.000000                        0.000000               0.7212              0.00000        0.22222                       0.77049                     0.000000           0.67816                0.7692            0.73684
Specificity                          0.995342         0.998464                        0.993769               0.8347              1.00000        0.96640                       0.96167                     0.995305           0.95470                0.8634            0.96581
Pos Pred Value                       0.000000         0.000000                        0.000000               0.5920                  NaN        0.27586                       0.67143                     0.000000           0.69412                0.6349            0.73684
Neg Pred Value                       0.974164         0.984848                        0.971081               0.9000              0.98185        0.95570                       0.97631                     0.966565           0.95139                0.9237            0.96581
Prevalence                           0.025719         0.015129                        0.028744               0.2496              0.01815        0.05446                       0.09228                     0.033283           0.13162                0.2360            0.11498
Detection Rate                       0.000000         0.000000                        0.000000               0.1800              0.00000        0.01210                       0.07110                     0.000000           0.08926                0.1815            0.08472
Detection Prevalence                 0.004539         0.001513                        0.006051               0.3041              0.00000        0.04387                       0.10590                     0.004539           0.12859                0.2859            0.11498
Balanced Accuracy                    0.497671         0.499232                        0.496885               0.7779              0.50000        0.59431                       0.86608                     0.497653           0.81643                0.8163            0.85133
</code></pre></div></p>
<p>Note that the accuracy is low-ish by design.
   We wanted a teaching data set we can demonstrate how to improve performance through tuning or by using different models or analysis techniques.</p>
<h2 id="documentation-of-data-set-contents">Documentation of data set contents</h2>
<h3 id="column-descriptions">Column descriptions</h3>
<ul>
<li><code>peptipedia_peptide_id</code>: peptide id of the peptide in the peptipedia database. All peptides were representative sequences in mmseqs2 clustering (80% threshold).</li>
<li><code>peptide_sequence</code>: peptide sequence.</li>
<li><code>bioactivity</code>: bioactivity label for the peptide. This was derived by taking all bioactivity labels for all peptides in a cluster and selecting only clusters where all peptides had at least one label that was assigned to all peptides. We kept the first of such labels and use this as the "bioactivity" label for the peptide.</li>
<li><code>all_bioactivities</code>: all predicted bioactivities for a given peptide.</li>
<li><code>go_id</code>: gene ontology ID.</li>
<li><code>pfam_id</code>: PFAM (protein family) ID.</li>
<li><code>pfam_name</code>: name of PFAM protein family.</li>
<li><code>pfam_type</code>: type of PFAM protein family.</li>
<li><code>aliphatic_index</code>: relative volume occupied by aliphatic side chains (Alanine, Valine, Isoleucine, and Leucine).</li>
<li><code>boman_index</code>: The potential interaction index proposed by Boman (2003) is an index computed by averaging the solubility values for all residues in a sequence. It can be used to give an overall estimate of the potential of a peptide to bind to membranes or other proteins. A value greater than 2.48 indicates that a protein has high binding potential.</li>
<li><code>charge</code>: the theoretical net charge of a peptide sequence based on the Henderson-Hasselbach equation. Computed at a pH of 7 with the Lehninger pKscale.</li>
<li><code>hydrophobicity</code>: average of the hydrophobicity values of each residue using one of the 39 scales from different sources.</li>
<li><code>instability_index</code>: predicts the stability of a protein based on its dipeptide composition. A protein whose instability index is smaller than 40 is predicted as stable, a value above 40 predicts that the protein may be unstable.</li>
<li><code>isoelectric_point</code>: the pH at which a particular molecule or surface carries no net electrical charge.</li>
<li><code>molecular_weight</code>: the sum of the mass of each amino acid in the peptide sequence.</li>
<li><code>pd1_residue_volume</code>: physical descriptor related to the peptide residue volume.</li>
<li><code>pd2_hydrophilicity</code>: physical descriptor related to the peptide hydrophilicity.</li>
<li><code>z1_lipophilicity</code>: Zscale quantifying the lipophilicity of the peptide.</li>
<li><code>z2_steric_bulk_or_polarizability</code>: Zscale modeling steric properties like steric bulk and polarizability.</li>
<li><code>z3_polarity_or_charge</code>: Zscale quantifying electronic properties like polarity and charge.</li>
<li><code>z4_electronegativity_etc</code>: Zscale relating to electronegativity, heat of formation, electrophilicity and hardness.</li>
<li><code>z5_electronegativity_etc</code>: Zscale relating to electronegativity, heat of formation, electrophilicity and hardness.</li>
<li><code>length</code>: length in amino acids of the peptide.</li>
</ul>
<h2 id="potential-future-additions-or-improvements">Potential future additions or improvements</h2>
<ol>
<li>We could make the data set creation script into a Snakemake workflow to better document how it was generated.</li>
<li>We could BLAST each peptide sequence to add more metadata (peptide function, gene name, species/taxonomy, number of hits, etc). This might be helpful especially if we need more character columns.</li>
</ol>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2022 <a href="https://www.arcadiascience.com">Arcadia Science</a>
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.5bf1dace.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.078830c0.min.js"></script>
      
    
  </body>
</html>